### PR TITLE
fix(build): switch to esbuild and tsc

### PR DESCRIPTION
**Pull request by kfcampbell**
Replicating [octokit/request.js#584](https://github.com/octokit/request.js/pull/584) to another repo. 

Consumers of this module such as [oauth-methods.js#152](https://github.com/octokit/oauth-methods.js/pull/152) are seeing missing types. This PR attempts to fix that issue by getting us off of pika before additional Node deprecations can be made. 

Note that when running `npm run build` locally, I see the following errors:

```
 sh$ npm run build

> @octokit/oauth-authorization-url@0.0.0-development build
> node scripts/build.mjs && tsc -p tsconfig.json

✘ [ERROR] Could not resolve "./types"

    pkg/dist-src/index.js:7:7:
      7 │ } from "./types";
        ╵        ~~~~~~~~~

✘ [ERROR] Could not resolve "./types"

    pkg/dist-src/index.js:7:7:
      7 │ } from "./types";
        ╵        ~~~~~~~~~
```

These errors are not present in the CI build because `npm run build` does not appear to be run then. I believe this to be a configuration issue and I think I missed something in this repo to get this to function correctly. Tips are appreciated!

----

Pika has been deprecated for a while now, and the project has now been archived, and because of that the released npm package is missing most of the files generated by the build step.

Uses esbuild to transpile the TS source code into an ESM source, NodeJs bundle, and a browser bundle

Uses the TypeScript compiler to generate the types.

### Before the change?

* Uses pika for the build system
* The released npm package is missing most of the files generated by the build step. `dist-node`, `dist-types`, `dist-web`... even though they are generated correctly.

### After the change?

* Uses esbuild to generate the JS bundles
* Uses the TypeScript compiler to generate the type definitions
* The correct file matching patterns are used so we publish all the necessary files again.

----

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/main/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:

* N/A

### Pull request type

Please add the corresponding label for change this PR introduces:
- Bugfix: `Type: Bug`
- Dependencies/code cleanup: `Type: Maintenance`

----

### CODE REVIEW

File: scripts/build.mjs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  const entryPoints = ["./pkg/dist-src/index.js"];
+
+  await Promise.all([
+    // Build the a CJS Node.js bundle
+    esbuild.build({
+      entryPoints,
+      outdir: "pkg/dist-node",
+      bundle: true,
+      platform: "node",
+      target: "node14",
```
Comment (wolfy1339) [inline scripts/build.mjs:45]:
```suggestion
      target: "node18",
```

File: src/index.ts [hunk 2]
```diff
@@ -1,5 +1,4 @@
 import {
```
Comment (wolfy1339) [inline src/index.ts:1]:
```suggestion
import type {
```

**Review by wolfy1339 — CHANGES_REQUESTED** [review]
Some quick things I noticed

**Comment by wolfy1339** [comment]
The errors are because you need to explicitly mark type imports/exports

**Comment by kfcampbell** [comment]
@wolfy1339 thanks for the tip! After applying your changes I'm still seeing the same errors...is there another spot where an import/export is missed?

**Comment by kfcampbell** [comment]
Aha! [Here](https://github.com/octokit/oauth-authorization-url.js/pull/279/commits/ab61e225d85ba07635675ba8b4a8707c9cb43a01) is what was missing. `npm run build` succeeds for me locally and I think this is ready to go.

File: scripts/build.mjs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        types: "dist-types/index.d.ts",
+        module: "dist-src/index.js",
+        sideEffects: false,
+      },
+      null,
+      2
+    )
+  );
+}
+main();
```
Comment (wolfy1339) [inline scripts/build.mjs:88]: Add new line at the end of the file

File: tsconfig.json [hunk 4]
```diff
@@ -1,9 +1,11 @@
...
-    "strict": true
+    "esModuleInterop": true,
+    "declaration": true,
+    "outDir": "pkg/dist-types",
+    "emitDeclarationOnly": true,
+    "sourceMap": true
   },
-  "include": ["src/index.ts"]
+  "include": ["src/**/*"]
 }
```
Comment (wolfy1339) [inline tsconfig.json:11]: Add newline at the end of the file

**Review by wolfy1339 — APPROVED** [review]
LGTM, some small style issues.

File: tsconfig.json (hunk 4 above)
Comment (kfcampbell) [reply tsconfig.json]: Done!

File: scripts/build.mjs (hunk 3 above)
Comment (kfcampbell) [reply scripts/build.mjs]: Done!

**Pull request merged**